In [13]:
import math

## 📘 Coding Challenge: Pearson Correlation Coefficient

### 🧠 Introduction

In this challenge, you'll implement the Pearson Correlation Coefficient using pure Python.
No numpy. No pandas. Just math and list comprehension magic.

This metric is used to measure the strength and direction of a linear relationship between two variables.

---





📐 Formula

$ 
r = \frac{\sum (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum (x_i - \bar{x})^2} \cdot \sqrt{\sum (y_i - \bar{y})^2}} 
$

Where:
- $ x_i $, $ y_i $ are the values
- $ \bar{x} $, $ \bar{y} $ are the means of the respective variables

In [2]:
# ✅ Starter Data
x = [1, 2, 3, 4, 5]
y = [2, 4, 5, 4, 5]

TODO: Step 1 - Calculate the means of x and y

In [3]:
mean_x = sum(x) / len(x)
mean_y = sum(y) / len(y)

TODO: Step 2 - Calculate the numerator: $ \text{sum of }(x_i - \bar{x}) * (y_i - \bar{y}) $

In [23]:
numerator = sum([(xi - mean_x) * (yi - mean_y) for xi, yi in zip(x, y)])
numerator

6.0

TODO: Step 3 - Calculate the denominator:

$
\sqrt{\sum (x_i - \bar{x})^2} \times \sqrt{\sum (y_i - \bar{y})^2}
$

In [25]:

denominator = math.sqrt(sum((xi - mean_x) **2 for xi in x)) * math.sqrt(sum((yi - mean_y) **2 for yi in y))
denominator

7.745966692414833

TODO: Step 4 - Calculate r = numerator / denominator

In [26]:
result = numerator / denominator
result

0.7745966692414834

Optional: Round to 4 decimal places for readability

In [27]:
round(result, 4)

0.7746

Expected: A value between -1 and 1 indicating correlation strength
Example output: r = 0.7746

Print it
print(f"Pearson Correlation Coefficient: {r}")

In [28]:
print(f"Pearson Correlation Coefficient: {round(result, 4)}")

Pearson Correlation Coefficient: 0.7746



🎯 Bonus Challenge
We are not going to do this, but it's just a thought to leave with you for when you are fealing fiesty! 🏑

Create a function `pearson_corr(x, y)` that returns r.

Test it with different input pairs.